src/devices/bus/epson_qx/sound_card.cpp

```diff
@@ -0,0 +1,127 @@
+
+DEFINE_DEVICE_TYPE(EPSON_QX_OPTION_YM2149, bus::epson_qx::ym2149_sound_card_device, "epson_qx_option_sound_card", "Epson QX-10 YM2149 Sound Card")
+
+namespace bus::epson_qx {
+
+static INPUT_PORTS_START( ym2149_sound_card )
+	PORT_START("IOBASE")
+	PORT_CONFNAME(0xf0, 0xc0, "IO Base Address Selection")
+	PORT_CONFSETTING(0x80, "&80")
+	PORT_CONFSETTING(0x90, "&90")
+	PORT_CONFSETTING(0xa0, "&A0")
+	PORT_CONFSETTING(0xb0, "&B0")
+	PORT_CONFSETTING(0xc0, "&C0")
+	PORT_CONFSETTING(0xd0, "&D0")
+	PORT_CONFSETTING(0xe0, "&E0")
+	PORT_CONFSETTING(0xf0, "&F0")
+	PORT_START("IRQ")
+	PORT_CONFNAME(0x03, 0x03, "MPU IRQ")
+	PORT_CONFSETTING(0x00, "NONE")
+	PORT_CONFSETTING(0x01, "INTH1")
+	PORT_CONFSETTING(0x02, "INTH2")
+	PORT_CONFSETTING(0x03, "INTL")
+INPUT_PORTS_END
+
+//-------------------------------------------------
+//  ym2149_sound_card_device - constructor
+//-------------------------------------------------
+ym2149_sound_card_device::ym2149_sound_card_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, EPSON_QX_OPTION_YM2149, tag, owner, clock),
+	device_option_expansion_interface(mconfig, *this),
+	m_mpu401(*this, "mpu401"),
+	m_ym2149(*this, "ym2149"),
+	m_iobase(*this, "IOBASE"),
+	m_irq(*this, "IRQ"),
+	m_installed(false)
+{
+}
+
+//-------------------------------------------------
+//  device_input_ports - device-specific ports
+//-------------------------------------------------
+ioport_constructor ym2149_sound_card_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME( ym2149_sound_card );
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+void ym2149_sound_card_device::device_add_mconfig(machine_config &config)
+{
+	MPU401(config, m_mpu401).irq_cb().set(FUNC(ym2149_sound_card_device::mpu_irq_out)).invert();
+	SPEAKER(config, "speaker").front_center();
+	YM2149(config, m_ym2149, clock()).add_route(ALL_OUTPUTS, "speaker", 1.0);
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+void ym2149_sound_card_device::device_start()
+{
+	m_installed = false;
+
+	save_item(NAME(m_installed));
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void ym2149_sound_card_device::device_reset()
+{
+	m_ym2149->set_pin26_low_w();
+
+	if (!m_installed) {
+		address_space &space = m_bus->iospace();
+		offs_t iobase = m_iobase->read() & 0xf0;
+		space.install_device(iobase, iobase+0x03, *this, &ym2149_sound_card_device::map);
+		m_installed = true;
+	}
+}
+
+uint8_t ym2149_sound_card_device::ym2149_read(offs_t offset)
+{
+	return m_ym2149->data_r();
+}
+
+void ym2149_sound_card_device::map(address_map &map)
+{
+	map(0x01, 0x01).r(FUNC(ym2149_sound_card_device::ym2149_read));
```

Comment:
I don't see the need for a trampoline here, should just work with `.m(m_ym2149, FUNC(ym2149_device::data_r));`

Also a suggestion: I find it more readable to read `m_ssg` rather than repeating device name with `m_ym2149`, also given the `ym2149_sound_card_device` itself.

Reply:
I've gone and removed the unnecessary trampoline, I didn't realize the delgate also accepted functions with no parameters.

Reply:
There’s a bunch of overload magic to allow the address space, offset, and access mask arguments to be omitted if they aren’t necessary, although the offset can only be omitted if the access mask is also omitted.